keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+#### Storing Logger and metadata in Context
+
+We would like to propose that logger interface and all metadata collected trough call stack to be stored in Golang Context object. This way creating one canonical way to pass contextual information across the call stack. Some pros of this approach:
+* Standard method in Golang to pass scoped variables across API boundaries and between processes.
+* No need for additional function argument for passing logging API through call stack
+* Context is used similarly by many frameworks including for instance OpenCensus tracing API.
+* Accessing metadata will be independent from Logging and could be reused by other functionalities.
+
+Access to context variables should be done though predefined functions. For that we propose to a new module to kubernetes component base called `k8s.io/component-base/kontext`. This new module will include all utility functions for storing and retrieving logger and metadata needed by it.
+
+Throughout call stack new metadata can be added to context that can be later used by logger. For example reconciliation loop can reference the reconcilided object so that all logs will be referencing it. All metadata access should be done through functions in `k8s.io/component-base/kontext` module. Adding reference to kubernetes object (pod, service, node) should be really straightforward.
+
+#### Example
+
+```go
+import "k8s.io/component-base/kontext"
+
+func UpdateStatus(ctx context.Context, pods []*v1.Pod, newStatus bool) {
+	for _, pod := range pods {
+		podCtx := kontext.WithObject(ctx, pod)
+		updatePodStatus(podCtx, pod, newStatus)
+	}
+}
+
+func updatePodStatus(ctx context.Context, pod *v1.Pod, newStatus bool) {
+  logger := kontext.GetLogger(ctx).WithName("probe")
+  logger.V(6).Info("Updating ready status", "ready", newStatus)
+  [...]
+  logger.V(4).Info("Ready status updated", "ready", newStatus)
+}
+```
+
+### New logging formats
+
+Introduction of new logging API should be paired with new output format that can properly utilize structured logging. In order to simplify log ingestion, but still maintain similar developer experience we would like to propose two formats. First one for production use and second one for component development.
+
+#### Json format
+
+We would like to propose a JSON as a future default format in Kubernetes for both production and development use. Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```json
+{
+   "ts":"00:15:15.525108",
+   "msg":"Updating ready status",
```

Comment:
I'd be interested to know if from a UX perspective we could guarantee some ordering of these values.

In the current logging scenario (glog/klog), the expectations are set, a user knows where to look for a message, or an error, etc. JSON is a good machine-readable format, less readable for humans. Do we expect users to lean towards other tooling to actually read logs?

For `kubectl logs`, have we thought about supporting something that can format in a human readable manner these structured output with a new flag or similar?

Reply:
In original proposal for this KEP apart from a json output format we had also a dev output format which would be something in between the current klog format and json format.
The usability of it has been questioned by a few people and therefore we have decided to remove it for the sake of simplicity.
When it comes to json output format we definitely would like to have consistent order of fields in them which would make easier not only to read those logs but also to process them by non json aware tools like for instance sed or awk.
When reading those logs by humans passing them through jq or similar tool should help a lot. Although we have not considered it before some special support for json formatted logs in `kubectl logs` might help also with the readability.

Reply:
> For kubectl logs, have we thought about supporting something that can format in a human readable manner these structured output with a new flag or similar?

+1 on the idea (longterm, probably not an initial detail) of having kubectl logs or some other tooling be structured aware.

> When it comes to json output format we definitely would like to have consistent order of fields in them which would make easier not only to read those logs but also to process them by non json aware tools like for instance sed or awk.

I'd imaging sorting *might* have negative performance impact, so it's probably best to measure that first before deciding what we want to do here.  Operations like sorting fields in ingestion are pretty trivial (`jq -S` for sorting).

As for awk and sed, I'm not actually certain it's really necessary to tailor to them -- you've got structured data, why try and parse with non-structured tooling except in certain edge cases?

> a user knows where to look for a message, or an error, etc

Kinda -- we're *comfortable* with k8s's klog error messages, but actually finding all the cases of error messages, or a particular thing are pretty hard, since we can change how errors are wrapped, where they appear in the message, etc between log lines.  It's mostly consistent, but only mostly.

Reply:
> When reading those logs by humans passing them through jq or similar tool should help a lot. 

Yes and no, it does help but it also creates an extra step. The suggestion of adding making `kubectl logs` structured-logging aware is one where we could, by knowing the structure of each line in advance, develop opinionated views of incoming logs: "show me every line with level => WARN", "filter my logs on condition X", "give me everything after a certain date", "show me only important bits in human readable format", and so on. 

> +1 on the idea (longterm, probably not an initial detail) of having kubectl logs or some other tooling be structured aware.

That's something I'd definitely like to add to the GA readiness checklist, definitely not for the first iteration.

> I'd imaging sorting might have negative performance impact, so it's probably best to measure that first before deciding what we want to do here. Operations like sorting fields in ingestion are pretty trivial (jq -S for sorting).

I should have clarified a little better what I meant, my suggestion was more to guarantee that some fields would always come at the top, like say "ts" = 1, "level" = 2, and  "msg" = 3, and so on. This behavior would solve the immediate issue of having logs act in a familiar way (like glog is structured), and help with general readability expectations.

Reply:
Teaching `kubectl logs` about the structure implies that the final log-lines are formatted in a particular way.  Concretely that means choosing glogr vs zapr vs ... which, in part, this KEP tries to avoid.

In theory it should be possible for a user to choose a totally different log lib.